community: add reka chat model integration

findalexli opened the pull request:
community: add Reka chat model integration

Description: This PR adds support for the Reka chat model to the LangChain community package. It includes the implementation of ChatReka class, which allows users to interact with Reka's API for chat completions. The integration supports both text and image inputs, streaming responses, and async operations.

Dependencies: This integration inside langchain-community requires the reka-api package to be installed. Users will need to install it separately with pip install reka-api.

Added features:

ChatReka class for interacting with Reka chat models
Support for text and image inputs
Streaming and non-streaming response handling
Synchronous and asynchronous methods
Unit and integration tests
Example notebook demonstrating usage
Testing:

Unit tests have been added in tests/unit_tests/chat_models/test_reka.py
Integration tests have been added in tests/integration_tests/chat_models/test_reka.py
An example notebook has been created in docs/docs/integrations/chat/reka.ipynb, including tool use with search. 
I have run make format, make lint, and make test from the root of the community package, and all checks pass.

X/Twitter handle: @RekaAILabs

findalexli committed:
Integration test

findalexli committed:
Linting fix

findalexli committed:
code formatter changes

findalexli committed:
Ruff linting check

findalexli committed:
Unit test update

findalexli committed:
Included dependecy

findalexli committed:
Fixed linting

findalexli committed:
Track changes to too lintegraiton

findalexli committed:
Merge branch 'langchain-ai:master' into master

findalexli committed:
Fix linting

findalexli committed:
Update notebook with tool use example

findalexli committed:
Merge branch 'master' into master

findalexli committed:
Tool use enfoce a version

efriis commented:
closing as duplicate of #26405 - let me know if they're different

efriis commented:
nvm - got it - personal branch

Commit:
Merge branch 'master' into master

efriis commented:
looks like CI might still be failing though

findalexli committed:
use pytest mark requires reka-api

findalexli committed:
Lint and mark test requring reka-api

findalexli committed:
Update reka import in notebook

findalexli committed:
Import order update to pass unit test

findalexli commented:
thanks. we pushed a few fix to address the CI issues.

 @efriis  we don't have the credentials to look into the vercel deployment issue, can you help with that?

findalexli committed:
Merge branch 'master' into master

vbarda committed:
Merge branch 'master' into master

vbarda committed:
lint

vbarda committed:
Merge branch 'master' into master

vbarda commented on libs/community/extended_testing_deps.txt:
```diff
@@ -82,7 +83,7 @@ sseclient-py>=1.8.0,<2
 sympy>=1.12,<2
 telethon>=1.28.5,<2
-tidb-vector>=0.0.3,<1.0.0
+tidb-vector>=0.0.3,<1.0.0   
```
```suggestion
tidb-vector>=0.0.3,<1.0.0
```

vbarda committed:
remove extra whitespace

vbarda commented on libs/community/langchain_community/chat_models/reka.py:
```diff
@@ -0,0 +1,411 @@
+from langchain_core.utils.function_calling import convert_to_openai_tool
+from pydantic import BaseModel, ConfigDict, Field, model_validator
+
+REKA_MODELS = [
```
this doesn't seem to be used anywhere

findalexli committed:
Remove unused model list

findalexli committed:
Merge branch 'master' into master

findalexli committed:
Merge branch 'master' into master

vbarda commented on libs/community/langchain_community/chat_models/reka.py:
```diff
@@ -0,0 +1,405 @@
+
+        # Ensure tool_choice is one of the allowed options
+        if tool_choice not in ("auto", "none", "tool"):
+            warnings.warn(
```
this should be an error instead

findalexli commented:
Our model is not trained with system prompt. Our api combines them just like our langchain integration.
On Oct 31, 2024 at 1:56 PM -0700, Vadym Barda ***@***.***>, wrote:
> @vbarda commented on this pull request.
> In libs/community/langchain_community/chat_models/reka.py:
> > +                if isinstance(message.content, str):
> +                    system_message = message.content
> +                else:
> +                    raise TypeError("SystemMessage content must be a string.")
> +            else:
> +                raise ValueError("Multiple system messages are not supported.")
> +        elif isinstance(message, HumanMessage):
> +            processed_content = process_content(message.content)
> +            if system_message:
> +                if (
> +                    processed_content
> +                    and isinstance(processed_content[0], dict)
> +                    and processed_content[0].get("type") == "text"
> +                    and "text" in processed_content[0]
> +                ):
> +                    processed_content[0]["text"] = (
> what is the reason for combining the system message with a human message? from your API ref here https://docs.reka.ai/api-reference/chat/create looks like role = system should be supported?
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***>

vbarda commented on libs/community/tests/unit_tests/chat_models/test_reka.py:
```diff
@@ -0,0 +1,210 @@
+        ),
+    ],
+)
+def test_convert_to_reka_messages(
```
i think this needs a few more tests for working with system messages

vbarda commented:
@findalexli thanks for the contribution! i added a few comments, apologies for incomplete review last week

vbarda commented on libs/community/langchain_community/chat_models/reka.py:
```diff
@@ -0,0 +1,405 @@
+                    and processed_content[0].get("type") == "text"
+                    and "text" in processed_content[0]
+                ):
+                    processed_content[0]["text"] = (
```
> Our model is not trained with system prompt. Our api combines them just like our langchain integration.

saw your response, threading here. if the API already handles system messages and converts them, why do this in the LangChain integration as well?

findalexli committed:
Combine reka chat into one class

findalexli committed:
Unit and integration test with system messages

findalexli committed:
Notebook doc update (pip, e2e agent)

findalexli committed:
Lint/formatt

findalexli commented:
Addressed the tests and a couple issues you pointed out, thank you @vbarda !

findalexli committed:
Merge branch 'master' into master

vbarda commented:
> Addressed the tests and a couple issues you pointed out, thank you @vbarda !

@findalexli thanks for adding more tests! there is still a bunch of comments that are not addressed, or marked resolved but don't seem to be resolved. could you please respond to those and update the PR where appropriate?

findalexli committed:
Linted notebook

findalexli committed:
add chatke in test import

findalexli committed:
add tiktoken token count method

findalexli commented:
HI @vbarda  we made some updates and have addressed all of your comments (including notebook, token count, duplicated class, system role).  Added a patch in the `test_import` that were the source failing the ci test. Run lint and format. Thanks!

vbarda commented on libs/community/langchain_community/chat_models/reka.py:
```diff
@@ -152,8 +154,9 @@ class RekaCommon(BaseLanguageModel):
-    count_tokens: Optional[Callable[[str], int]] = None
     model_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    model_config = ConfigDict(extra="forbid")
+    _tiktoken_encoder = None
```
why is this private? could we make it into a more general `token_counter` param with better types? something like this https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/messages/utils.py#L577

vbarda commented on libs/community/langchain_community/chat_models/reka.py:
```diff
@@ -339,11 +331,9 @@ async def _agenerate(
-                "get_num_tokens() is not implemented for Reka models."
-            )
-        return self.count_tokens(text)
+        if self._tiktoken_encoder is None:
```
suggest making this into a more generic `token_counter` attribute that's not private

vbarda commented:
> HI @vbarda we made some updates and have addressed all of your comments (including notebook, token count, duplicated class, system role). Added a patch in the `test_import` that were the source failing the ci test. Run lint and format. Thanks!

Thanks! Looks like there are a few things failing (due to missing `tiktoken` dependency)

findalexli committed:
Add required header in doc notebook

findalexli committed:
token count update

findalexli commented:
Updated to address the token count comment. thanks for the feedback. I have ran make lint, make lint and make test, and make extended_tests which passed.

vbarda commented on libs/community/langchain_community/chat_models/reka.py:
```diff
@@ -329,11 +330,29 @@ async def _agenerate(
         return ChatResult(generations=[ChatGeneration(message=message)])
 
-    def get_num_tokens(self, text: str) -> int:
+    def get_num_tokens(self, input: Union[str, BaseMessage, List[BaseMessage]]) -> int:
```
looks like the input type here is not matching the one in the `token_counter` callable (failing linter)

vbarda reviewed (commented):
looks good to me -- thanks for addressing! i think one lint issue remaining and should be good to merge

findalexli committed:
Type consistency for token counter

vbarda committed:
Merge branch 'master' into master

vbarda committed:
Merge branch 'master' into master

vbarda committed:
remove from deps

vbarda committed:
remove scheduled

vbarda committed:
skip

vbarda committed:
lint

vbarda committed:
skip

vbarda committed:
update

vbarda committed:
fix link

vbarda committed:
Merge branch 'master' into master

vbarda committed:
update concepts link
